src/ergtable.h

```diff
@@ -0,0 +1,188 @@
+#ifndef ERGTABLE_H
+#define ERGTABLE_H
+
+#include <QList>
+#include <QSettings>
+#include <QObject>
+#include <QDebug>
+#include <QDateTime>>
+#include "qzsettings.h"
+
+struct ergDataPoint {
+    uint16_t cadence = 0; // RPM
+    uint16_t wattage = 0; // Watts
+    uint16_t resistance = 0; // Some unit
+
+    ergDataPoint() = default;
+
+    ergDataPoint(uint16_t c, uint16_t w, uint16_t r) : cadence(c), wattage(w), resistance(r) {}
+};
+
+Q_DECLARE_METATYPE(ergDataPoint)
+
+class ergTable : public QObject {
+    Q_OBJECT
+
+public:
+    ergTable(QObject *parent = nullptr) : QObject(parent) {
+        loadSettings();
+    }
+
+    ~ergTable() {
+        saveSettings();
+    }
+
+    void collectData(uint16_t cadence, uint16_t wattage, uint16_t resistance, bool ignoreResistanceTiming = false) {
+        if(resistance != lastResistanceValue) {
+            qDebug() << "resistance changed";
+            lastResistanceTime = QDateTime::currentDateTime();
+            lastResistanceValue = resistance;
+        }
+        if(lastResistanceTime.msecsTo(QDateTime::currentDateTime()) < 1000 && ignoreResistanceTiming == false) {
+            qDebug() << "skipping collecting data due to resistance changing too fast";
+            return;
+        }
```

Comment:
What if the cadence has changed but the resistance hasn't. This would throw away samples.

Also, I wonder if the wattage for a specific (cadence, resistance) would ever be different, e.g. if you were reading cadence from a cadence meter strapped to a pedal and power from a power meter inside the pedal.

Reply:
> What if the cadence has changed but the resistance hasn't. This would throw away samples.

the purpose of this is to thrown away every samples after a resistance change for 1 second. what issue do you see in this?

> Also, I wonder if the wattage for a specific (cadence, resistance) would ever be different, e.g. if you were reading cadence from a cadence meter strapped to a pedal and power from a power meter inside the pedal, rather than the bike's computer calculating it from its own measurement of cadence and what its resistance level achieves.


here i'm evaluating the power from the bike or from the power pedal based on the user configuration. so it's transparent

Reply:
> > What if the cadence has changed but the resistance hasn't. This would throw away samples.
> 
> the purpose of this is to thrown away every samples after a resistance change for 1 second. what issue do you see in this?

Based on the qDebug comment, I think you're looking to avoid collecting samples while the resistance is changing quickly.

It's currently waiting for 1000ms of stable resistance before storing the samples but could ignore changing cadence and wattage while the resistance remains stable for 1000ms. 

Also, if the user gradually decreased cadence, causing the resistance to increase at a rate of >1/s to match a target power, you could lose seconds of samples along the (C,R) line. This might actually be quite hard to do in practice though, with differing difficulty based on how many resistance levels the bike offers (i.e. you might have more "stability" as defined by this metric, from a bike that offers 16 resistance levels, over one that offers 250).

Reply:
> Based on the qDebug comment, I think you're looking to avoid collecting samples while the resistance is changing quickly.
> 
> It's currently waiting for 1000ms of stable resistance before storing the samples but could ignore changing cadence and wattage while the resistance remains stable for 1000ms.

Yes that's the purpose because some bikes could send resistance level changed without recalculate the wattage 


> Also, if the user gradually decreased cadence, causing the resistance to increase at a rate of >1/s to match a target power, you could lose seconds of samples along the (C,R) line. This might actually be quite hard to do in practice though, with differing difficulty based on how many resistance levels the bike offers (i.e. you might have more "stability" as defined by this metric, from a bike that offers 16 resistance levels, over one that offers 250).

Yes I considered this but I guess it's more important to don't add some spurious samples than adding more samples  :)

Reply:
> here i'm evaluating the power from the bike or from the power pedal based on the user configuration. so it's transparent

If the bike provides the cadence and power level all itself, it could be capable of working it out from its own internal table of (C,R,W), and always give you the same W for the (C,R). However, if the C and W are coming from independent sources that don't know the R, their individual margins of error may lead to a range of W for any given (C,R). So if the first (C,R,W) you collect has a sporadically bad W value, the current implementation is stuck with it.

Reply:
this scenario is impossible because i'm resetting the table if the user change the power sensor setting. so the table is or only filled with bike metrics or only filled with power pedal one + resistance